lib/dockerfile.ts

```diff
@@ -12,267 +12,184 @@ limitations under the License.
 import * as _ from 'lodash';
-import * as minimatch from 'minimatch';
 import * as path from 'path';
 
+import ActionGroup from './action-group';
 import { DockerfileParseError, UnsupportedError } from './errors';
+import Stage from './stage';
 
-/**
- * This interface represents a file which when
- * changed causes an option group to retrigger
- */
-export interface FileDependency {
-	/**
-	 * The path of the file on disk
-	 */
-	localPath: string;
-	/**
-	 * Where this file gets mapped to inside the container
-	 */
-	containerPath: string;
-	/**
-	 * Does the container path point to a directory or a file?
-	 */
-	destinationIsDirectory: boolean;
-	/**
-	 * Does the source path point to a directory or file?
-	 */
-	sourceIsDirectory: boolean;
+interface StagedActionGroups {
+	[stage: number]: ActionGroup[];
 }
 
-export type Command = string;
+export class Dockerfile {
+	public stages: Stage[] = [];
 
-/**
- * This interface represents a set of actions to be performed
- * in the same working directory on the container. They are
- * generated given the Dockerfile and a list of changed files.
- */
-export interface DockerfileActionGroup {
-	/**
-	 * A list of commands to run in order
-	 */
-	commands: Command[];
-	/**
-	 * A list of files which this group depends on.
-	 * What this means is that if any of the files in
-	 * this structure changes, every command in the `commands`
-	 * field should be re-ran
-	 */
-	fileDependencies: FileDependency[];
-	/**
-	 * The directory in the container which these commands should be
-	 * run in
-	 */
-	workDir: string;
-}
+	public constructor(dockerfileContent: string | Buffer) {
+		this.parse(dockerfileContent.toString());
+	}
 
-class Dockerfile {
-	public actionGroups: DockerfileActionGroup[] = [];
+	public getActionGroupsFromChangedFiles(files: string[]): StagedActionGroups {
+		// go through stage by stage, detecting which stages would be affected by
+		// this file being changed
+		files = files.map(path.normalize);
+		const stagedGroups: StagedActionGroups = {};
+		for (const [idx, stage] of this.stages.entries()) {
+			const actionGroups = stage.getActionGroupsForChangedFiles(files);
+			if (actionGroups.length > 0) {
+				stagedGroups[idx] = actionGroups;
+			}
+		}
 
-	public constructor(dockerfileContents: string) {
-		this.parse(dockerfileContents);
+		// Recursively detect changes in stages. We start with the stages
+		// which change as a direct result of a local copy. We then move
+		// onto stages which have become invalidated due to a change in
+		// a stage that they depend on. We then move onto stages which have
+		// been invalidated due to the stages invalidated by other stages,
+		// ad infinitum. We should probably add in infinite loop detection,
+		// but in general, to get to this point docker would have had to
+		// build the dockerfile anyway, so we can assume all is good.
+		const buildDependencyGraph = ([stageIdx, ...tail]: number[]): void => {
```

Comment:
Interesting. Is this an instance of [topological sorting](https://en.wikipedia.org/wiki/Topological_sorting)? _("Scheduling a sequence of jobs or tasks based on their dependencies")_ Just thinking if we would benefit from using some off-the-shelf npm package for that:
* https://www.npmjs.com/package/toposort
* https://www.npmjs.com/package/topo
* https://www.npmjs.com/search?q=keywords%3Atopological&ranking=popularity

But the custom implementation looks simple enough, and looks like if it's been well tested. 👍 
(And I think we can assume that the input is fairly small in size, so "big-O complexity" is probably not an issue.)